File: C++/BFS.cpp
#include <iostream>
#include <list>
#include <queue>
using namespace std;

class Graph {
    int V;
    list<int>* adj;
public:
    Graph(int V);
    void addEdge(int v, int w);
    void BFS(int s);
};

Graph::Graph(int V) {
    this->V = V;
    adj = new list<int>[V];
}

void Graph::addEdge(int v, int w) {
    adj[v].push_back(w);
}

void Graph::BFS(int s) {
    bool* visited = new bool[V];
    fill(visited, visited + V, false);
    
    queue<int> q;
    visited[s] = true;
    q.push(s);
    
    while (!q.empty()) {
        s = q.front();
        q.pop();
        cout << s << " ";
        
        for (int neighbor : adj[s]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }
    delete[] visited;
}

int main() {
    Graph g(4);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2);
    g.addEdge(2, 0);
    g.addEdge(2, 3);
    g.addEdge(3, 3);

    cout << "Breadth-First Search (starting from vertex 2): ";
    g.BFS(2);

    return 0;
}

File: PHP/BFS.php
<?php
function bfs($graph, $start) {
    $visited = [];
    $queue = [$start]; // Initialize queue with the starting node

    while (!empty($queue)) {
        $node = array_shift($queue); // Dequeue a vertex from queue
        if (!in_array($node, $visited)) {
            $visited[] = $node; // Mark the node as visited

            // Enqueue all adjacent vertices that have not been visited
            foreach ($graph[$node] as $neighbor) {
                if (!in_array($neighbor, $visited)) {
                    $queue[] = $neighbor;
                }
            }
        }
    }

    return $visited;
}

// Example graph represented as an adjacency list
$graph = [
    'A' => ['B', 'C'],
    'B' => ['A', 'D', 'E'],
    'C' => ['A', 'F', 'G'],
    'D' => ['B'],
    'E' => ['B', 'H'],
    'F' => ['C'],
    'G' => ['C'],
    'H' => ['E']
];

$result = bfs($graph, 'A');
print_r($result); // Output: Array ( [0] => A [1] => B [2] => C [3] => D [4] => E [5] => F [6] => G [7] => H )
?>